File: USACO/2015-2016/Jan 2016/Gold/lightsout.java
// Lights Out

import java.util.*;
import java.io.*;

public class lightsout {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());
		
		int[] x = new int[N], y = new int[N];
		for(int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			x[i] = Integer.parseInt(st.nextToken());
			y[i] = Integer.parseInt(st.nextToken());
		}
		br.close();
		
		int[] len = new int[N];
		for(int i = 0; i < N; i++) len[i] = Math.abs(x[i] - x[(i+1)%N]) + Math.abs(y[i] - y[(i+1)%N]);
		
		int[] cost = new int[N+1];
		for(int i = 0; i < N-1; i++) cost[i+1] = cost[i] + len[i];
		
		for(int i = N-1; i > 0; i--) cost[i] = Math.min(cost[i], cost[i+1] + len[i]);
		
		int[] turn = new int[N];
		for(int i = 0; i < N; i++) {
			int v1 = (i-1+N)%N;
			int v2 = (i+1+N)%N;
			
			if((x[i] - x[v1]) * (y[i] - y[v2]) - (y[i] - y[v1]) * (x[i] - x[v2]) > 0) turn[i] = 1;
		}
		
		
		int result = 0;
		
		for(int start = 1; start < N; start++) {
			for(int fake = 1; fake < N; fake++) { // checking if this could possibly be the starting location
				if(start == fake) continue;
				int s = start, f = fake;
				int total = 0;
				while(s < N && f < N) {
					if(turn[s] != turn[f]) break;
					total += len[s];
					if(len[s] != len[f]) {
						s++;
						f++;
						break;
					}
					
					s++;
					f++;
				}
				
				s %= N; 
				f %= N;
				total += cost[s];
				result = Math.max(result, total - cost[start]);
			}
		}
		
		System.out.println(result);
	}
}

File: USACO/2015-2016/Jan 2016/Gold/lightsout.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <set>

using namespace std;

int N;
int x[205], y[205];
int cost[205];

int main (){
	cin >> N;
	for(int i = 0; i < N; i++) cin >> x[i] >> y[i];
	
	// -1 for clockwise, -2 for counterclockwise
	vector<int> S; //?
	S.push_back(0);
	for(int i = 0; i < N; i++) {
		int j = (i+1)%N;
		S.push_back(abs(y[i] - y[j]) + abs(x[i] - x[j]));
		
		int k = (j+1)%N;
		if((x[i] - x[j]) * (y[k] - y[j]) - (x[k] - x[j]) * (y[i] - y[j]) > 0)// right turn 
			S.push_back(-1);
		else
			S.push_back(-2);
	}
	S.back() = 0;
	
	for(int i = 0; i < N-1; i++)
		cost[i+1] = cost[i] + S[i*2+1];
	
	for(int i = N-1; i > 0; i--)
		cost[i] = min(cost[i], cost[i+1] + S[2*i+1]);
		
	// for(int i = 0; i <= N; i++) cout << cost[i] << endl;
		
	multiset<vector<int>> st;
	for(int i = 0; i < S.size(); i+=2) {
		for(int j = i+1; j <= S.size(); j+=2) {
			vector <int > v = vector<int>(S.begin() + i, S.begin() + j);
			//for(int i: v) cout << i << " ";
			//cout << endl;
			st.insert(vector<int>(S.begin() + i, S.begin() + j));
		}
	}
	//cout << endl;
	
	int result = 0;
	for(int i = 2; i < N; i+=2) {
		int c = 0;
		int j = i+1;
		for(; j <= S.size(); j+=2) {
			vector<int> v = vector<int>(S.begin()+i, S.begin() + j);
			//for(int a: v) cout << a << " ";
			if(st.count(vector<int>(S.begin()+i, S.begin() + j)) == 1) break;
			c += S[j];
			//cout << c << endl;
		}
		result = max(result, c + cost[j/2] - cost[i/2]);
	}
	cout << result << endl;
}
